specs/core/1_new_shards.md

```diff
@@ -0,0 +1,583 @@
+| `DOMAIN_SHARD_PROPOSAL` | `193` | |
+
+## Containers
+
+### `ShardBlockWrapper`
+
+_Wrapper for being broadcasted over the network._
+
+```python
+class ShardBlockWrapper(Container):
+    shard_parent_root: Hash
+    beacon_parent_root: Hash
+    slot: Slot
+    body: BytesN[SHARD_BLOCK_CHUNK_SIZE]
+    signature: BLSSignature
+```
+
+### `ShardSignableHeader`
+
+```python
+class ShardSignableHeader(Container):
+    shard_parent_root: Hash
+    beacon_parent_root: Hash
+    slot: Slot
+    body_root: Hash
+```
+
+### `ShardState`
+
+```python
+class ShardState(Container):
+    slot: Slot
+    gasprice: Gwei
+    data: Hash
+    latest_block_root: Hash
+```
+
+### `AttestationData`
+
+```python
+class AttestationData(Container):
+    slot: Slot
+    index: CommitteeIndex
+    # LMD GHOST vote
+    beacon_block_root: Hash
+    # FFG vote
+    source: Checkpoint
+    target: Checkpoint
+    # Current-slot shard block root
+    head_shard_root: Hash
+    # Shard transition root
+    shard_transition_root: Hash
+```
+
+### `ShardTransition`
+
+```python
+class ShardTransition(Container):
+    # Starting from slot
+    start_slot: Slot
+    # Shard block lengths
+    shard_block_lengths: List[uint64, MAX_SHARD_BLOCKS_PER_ATTESTATION]
+    # Shard data roots
+    shard_data_roots: List[List[Hash, MAX_SHARD_BLOCK_CHUNKS], MAX_SHARD_BLOCKS_PER_ATTESTATION]
+    # Intermediate shard states
+    shard_states: List[ShardState, MAX_SHARD_BLOCKS_PER_ATTESTATION]
+    # Proposer signature aggregate
+    proposer_signature_aggregate: BLSSignature
+```
+
+### `Attestation`
+
+```python
+class Attestation(Container):
+    aggregation_bits: Bitlist[MAX_VALIDATORS_PER_COMMITTEE]
+    data: AttestationData
+    custody_bits: List[Bitlist[MAX_VALIDATORS_PER_COMMITTEE], MAX_SHARD_BLOCKS_PER_ATTESTATION]
+    signature: BLSSignature
+```
+
+### `AttestationAndCommittee`
+
+```python
+class AttestationAndCommittee(Container):
+    committee: List[ValidatorIndex, MAX_VALIDATORS_PER_COMMITTEE]
+    attestation: Attestation
+```
+
+### `CompactCommittee`
+
+```python
+class CompactCommittee(Container):
+    pubkeys: List[BLSPubkey, MAX_VALIDATORS_PER_COMMITTEE]
+    compact_validators: List[uint64, MAX_VALIDATORS_PER_COMMITTEE]
+```
+
+### `AttestationCustodyBitWrapper`
+
+```python
+class AttestationCustodyBitWrapper(Container):
+    attestation_root: Hash
+    block_index: uint64
+    bit: bool
+```
+
+### `PendingAttestation`
+
+```python
+class PendingAttestation(Container):
+    aggregation_bits: Bitlist[MAX_VALIDATORS_PER_COMMITTEE]
+    data: AttestationData
+    inclusion_delay: Slot
+    proposer_index: ValidatorIndex
+    crosslink_success: bool
+```
+
+## Helpers
+
+### `get_online_validators`
+
+```python
+def get_online_indices(state: BeaconState) -> Set[ValidatorIndex]:
+    active_validators = get_active_validator_indices(state, get_current_epoch(state))
+    return set([i for i in active_validators if state.online_countdown[i] != 0])
```

Comment:
Is `set` really necessary?  Trying to optimize from client implementation's perspective

Reply:
It's not, but when things are unordered and unique we declare it as `Set` in the spec. Here it would be a free conversion for a client to a set object anyway (or any validator indices container).